### ISSUE TITLE
2025/03/19: Intro to Transformers via a coding session

### ISSUE THREAD

**Issue by DuncDennis** [Dynamics-of-Neural-Systems-Lab/Generative-modelling-reading-group#46]
Hi, 

as a gentle introduction to Transformers I will lead a coding session next week, where we will code and train a Transformer in PyTorch. 
As a reference I will use this (amazing) tutorial by Andrej Karpathy: https://youtu.be/kCc8FmEb1nY?si=YEI2aWN6_fI3TFjh
The Transformer will be trained to auto-regressively predict the next character in a sequence, based on some training text. 

As you probably know, the Transformer architecture was first presented in the [Attention is all you need - paper](https://arxiv.org/abs/1706.03762). This paper is originally a machine-translation paper, meaning it also includes a cross-attention layer to map between - let's say - languages. In out coding session, we will only have the self-attention layer for auto-regressive forecasting. 

Notebook will follow.

**Comment by moritzschaefer** [comment]
The attention is all you need paper is really nicely written so it should provide great background for the next session!

**Comment by moritzschaefer** [comment]
Thanks for the great session. Here some of the questions that came up (feel free everyone to comment/answer them):

- Check what needs to be done to use transformers on longer sequences in generation/inference compared to training
- Why do Transformers not fine-tune well when unfreezing the last layers?
- Explain the normalization factor in front of the softmax? (by @seasonsOfTheSun)

**Comment by seasonsOfTheSun** [comment]
Thanks Dennis, that is the right answer!
